gradio/chat_interface.py

```diff
@@ -397,67 +403,86 @@ def _display_input(
         history.append([message, None])
         return history, history
 
-    def _submit_fn(
+    async def _submit_fn(
         self,
         message: str,
         history_with_input: list[list[str | None]],
         *args,
         **kwargs,
     ) -> tuple[list[list[str | None]], list[list[str | None]]]:
         history = history_with_input[:-1]
-        response = self.fn(message, history, *args, **kwargs)
+        if self.is_async:
+            response = await self.fn(message, history, *args, **kwargs)
+        else:
+            response = self.fn(message, history, *args, **kwargs)
```

Comment:
We need to run the non-coroutine in a separate thread otherwise we'll block the event loop. This is because the _submit_fn is now async.

If you run this code on your branch, two events can't run concurrently anymore

```python
import random
import gradio as gr
import time

def random_response(message, history):
    time.sleep(5)
    return random.choice(["Yes", "No"])

demo = gr.ChatInterface(random_response)

if __name__ == "__main__":
    demo.launch()
```

Note how the textbox submit event has to wait:
![chatbot_concurrency](https://github.com/gradio-app/gradio/assets/41651716/29103b89-2321-4b01-a5ec-91ff2436040e)

compare to main:
![chatbot_concurrency_main](https://github.com/gradio-app/gradio/assets/41651716/b7afcef0-5094-4d97-9d35-cc39729166ed)

Reply:
Ah nice thanks @freddyaboulton